File: 368-largest-divisible-subset/largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        // 09.02.24 potd
        int n=nums.size(), mx=1, num=-1;
        vector<int> ans, dp(n,1);
        sort(nums.begin(), nums.end());
        
        for(int i=1; i<n; i++){
            for(int j=0; j<i; j++){
                if(!(nums[i]%nums[j]) && dp[i]<dp[j]+1){
                    dp[i]=dp[j]+1;
                    
                    if(mx<dp[i])
                        mx=dp[i];
                }
            }
        }

        for(int i=n-1; i>=0; i--){
            if(mx==dp[i] && (num==-1 || !(num%nums[i]))){
                ans.push_back(nums[i]);
                mx--;
                num=nums[i];
            }
        }

        return ans;
    }
};

File: 368-largest-divisible-subset/largest-divisible-subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        # 06.04.25 potd
        ans = []
        nums.sort()
        dp = [1] * len(nums)
        prev = [-1] * len(nums)
        maxi = 0

        for i in range(1, len(nums)):
            for j in range(i):
                if nums[i]%nums[j] == 0 and dp[i] < dp[j] + 1:
                    dp[i] = dp[j] + 1
                    prev[i] = j
            
            if dp[i] > dp[maxi]:
                maxi = i
        
        idx = maxi
        while idx >= 0:
            ans.append(nums[idx])
            idx = prev[idx]
        
        return ans